### PR TITLE
feat(canonical): Add a shim for canonical keys

### PR DESCRIPTION
**Pull request by jan-auer**
This PR migrates the legacy "sentry.interfaces.*" event data keys to their new canonical representation, i.e. short names. Only canonical names will be stored, but access is still possible via the legacy names in the entire codebase. This includes ingestion and processing. 

A full list of renames is attached at the bottom.

I tried to keep this PR to a minimum, with more refactoring to happen afterwards. Especially EventManager code needs some serious clean up and I'd like to walk through the code base to remove every mention of "sentry.interfaces.*".

## Mapping Types

The conversion is mostly handled by two shim types:

 - `CanonicalKeyDict`: A mutable mapping based a dict that will only store canonical keys internally. Its constructor already converts and deduplicates all legacy keys. It does not synchronize with its input data.
 - `CanonicalKeyView`: A read-only view on an underlying mapping. When iterating or accessing keys, it internally checks for the legacy name as well as the canonical name in both directions. Additionally, it guarantees to preserve order for ordered data structures. Length and iteration operate on deduplicated canonical keys. For performance reasons, `__len__` is computed in the constructor, all other methods will operate on the underlying data.

Both wrappers prefer the canonical key in case both canonical and legacy keys are present in the input, regardless of the order. This also holds for OrderedDicts.

## Event Model Changes

The event model has been changed to now have two properties:

 - `event.node_data`: An instance of `NodeData` that was originally stored at `event.data`. It contains the raw mutable data dict. Apart from event creation and the _bind_nodes_ utility, it should never be used directly.
- `event.data`: A memoized read-only `CanonicalKeyView` on _node_data_. Since this is used by most of the Sentry code base, including plugins that operate on Event models, they continue to work with their legacy keys. Mutation of the event data should not have been allowed before, with the only exception of the _EventManager_ itself.

**Warning**: Since `event.data` is memoized, all calls to `Event.objects.bind_nodes(event, 'node_data')` **must** be made before the first access of `event.objects`. Otherwise, `len(event)` will return an invalid value. At the moment, this is the case in the entire codebase.

## API and EventManager Changes

The _EventManager_ used to rely on mutation of the data dict passed into its constructor. This is no longer the case, with the caveat that one **must** use the return value of `normalize` now. Apart from tests, there was only one affected location inside `LazyData` which was already updated in https://github.com/getsentry/sentry/pull/8774. 

```python
manager = EventManager(data)
data = manager.normalize()
```

The EventManager now uses a mutable `CanonicalKeyDict` to store data internally and returns it from _normalize_. This value is assumed directly by LazyData and thus carries through the entire store API code, allowing it and plugins to continue using legacy keys. Finally, the EventManager sets it directly into the Event's NodeData. This is the only place where `node_data` does **not** contain a raw dict.

Also, some code like `get_path` or the schema validation had to be updated to accommodate a non-dict data, either by using `isinstance(data, collections.Mapping)` or by accessing `data.data` (the internal raw dict).

## Processing and Store Changes

These jobs load event data without the Event model from the cache. Their entry points have been updated to wrap the returned data dict in a mutable `CanonicalKeyDict`. Since none of these jobs operates on Event models, no further changes were necessary.

Note that data no longer passes `isinstance(data, dict)`.

**Warning**: To ensure a smooth rollout, save and processing workers should be deployed before web. Otherwise, incompatible data with canonical keys might reach old workers.

## Next Steps

I'd not like this to go out right away because it has a high chance of breaking a lot. Ideally, we could deploy a small number of workers first before making the full change.

 - [ ] Validate assumptions and implications described above 
 - [ ] Verify that plugins continue to work
 - [ ] Deploy save / process workers
 - [ ] Deploy web
 - [ ] **Follow-up**: Clean up / refactor all uses of _"sentry.interfaces.*"_

## Renamed Keys

For reference, this is the mapping from legacy keys to canonical keys:

- `sentry.interfaces.Exception` -> `exception`,
- `sentry.interfaces.Message` -> `logentry`,
- `sentry.interfaces.Stacktrace` -> `stacktrace`,
- `sentry.interfaces.Template` -> `template`,
- `sentry.interfaces.Query` -> `query`,
- `sentry.interfaces.Http` -> `request`,
- `sentry.interfaces.User` -> `user`,
- `sentry.interfaces.Csp` -> `csp`,
- `sentry.interfaces.AppleCrashReport` -> `applecrashreport`,
- `sentry.interfaces.Breadcrumbs` -> `breadcrumbs`,
- `sentry.interfaces.Contexts` -> `contexts`,
- `sentry.interfaces.Threads` -> `threads`,
- `sentry.interfaces.DebugMeta` -> `debug_meta`,

### CODE REVIEW

**Comment by jan-auer** [comment]
@mattrobenolt @mitsuhiko I know it's kind of messy. If you're looking for a place to start reviewing, have a look at coreapi, EventManager, canonical.py, or store.py. Please also double-check if my view code matches your expectations of performance.

File: src/sentry/utils/samples.py
Comment (mitsuhiko) [inline src/sentry/utils/samples.py]: It this a change that creeped in or does this PR change behavior WRT to values?

File: src/sentry/api/endpoints/event_apple_crash_report.py
Comment (mitsuhiko) [inline src/sentry/api/endpoints/event_apple_crash_report.py]: This is going to be an issue. I much rather change `bind_nodes` so that it accepts the db column in addition to the descriptor and not change all calls to bind_nodes. Otherwise we need to touch some other stuff as well (hipchat-ac plugin though we can delete that, getsentry commands).

File: src/sentry/models/event.py
Comment (mitsuhiko) [inline src/sentry/models/event.py]: note for @mattrobenolt: the comment is a lie. Cannot be removed yet.

File: src/sentry/models/rawevent.py
Comment (mitsuhiko) [inline src/sentry/models/rawevent.py]: to mirror the comment above: we might want to keep the field under the original name since it affects `bind_nodes`. Additionally if we do this change I believe we need to reflect this in a dummy db migration.

File: tests/sentry/test_event_manager.py
Comment (mitsuhiko) [inline tests/sentry/test_event_manager.py]: Why is `keys()` necessary here? This is an odd statement and if it has a purpose a comment should clarify what it checks against the new list instead of the container itself.

File: src/sentry/utils/samples.py
Comment (jan-auer) [reply src/sentry/utils/samples.py]: That was a defensive change to keep behavior of the cocoa ingestion test the same. `cocoa.json` contains a key `breadcrumbs: [ ... ]`, which was not touched by the old code and would error with the new one. So instead of fixing this code, I chose to rather keep the test behavior the same.

**Comment by jan-auer** [comment]
This diff shows an example where the old undefined behavior and the new defined behavior are different: The example JSON contains both "user" and "sentry.interfaces.User": https://percy.io/getsentry/sentry/builds/804476

File: tests/sentry/test_event_manager.py
Comment (jan-auer) [reply tests/sentry/test_event_manager.py]: Because `__contains__` will say true if you ask `data` directly.

File: src/sentry/api/endpoints/event_apple_crash_report.py
Comment (jan-auer) [reply src/sentry/api/endpoints/event_apple_crash_report.py]: Can you clarify what you mean with db column vs descriptor please?

File: src/sentry/models/event.py
Comment (jan-auer) [reply src/sentry/models/event.py]: Should I remove it? cc @dcramer - right now the handling of `event.message` is quite messy and hard to understand. @mitsuhiko and I were thinking about cleaning this up soon, but we'd need some input on which way to go.

File: src/sentry/api/endpoints/event_apple_crash_report.py
Comment (jan-auer) [reply src/sentry/api/endpoints/event_apple_crash_report.py]: Reading the other comment, I think I get it now. That would mean that everyone accessing data would have to change to a new field then, instead of continuing to use `event.data`. I'm afraid that'd be a higher risk of breaking something. 

We could probably also patch this in `Event.objects.bind_nodes` and add a special case.

Comment (mitsuhiko) [reply src/sentry/api/endpoints/event_apple_crash_report.py]: @jan-auer i was thinking of letting `bind_nodes` also take the db_column into account as a fallback.

File: src/sentry/utils/samples.py
Comment (mitsuhiko) [reply src/sentry/utils/samples.py]: Is the cocoa dump the only one that does not have canonical data for value use? If that's the case I rather have the data fixed than special handle this here.

File: tests/sentry/test_event_manager.py
Comment (mitsuhiko) [reply tests/sentry/test_event_manager.py]: Can you add a comment clarifying this? This will likely to be picked up by a linter sooner or later and someone is going to "fix" it.

File: src/sentry/utils/samples.py
Comment (jan-auer) [reply src/sentry/utils/samples.py]: Sure thing. Will change the visual diff though.

File: src/sentry/models/rawevent.py
Comment (jan-auer) [reply src/sentry/models/rawevent.py]: Indeed, forgot to run migrations. I'll wait on a decision whether we should rename this or not. Regarding `bind_nodes`: If we do not rename this field, we'll have to come up with a new name for the accessor attribute and change every access to it. If we keep renaming it to `node_data` but don't want to change calls to bind_nodes, we'd have to look up the model's attribute name from the db_column (i.e. 'data' -> 'node_data'). There's probably a mapping somewhere in the meta model, but it sounds a bit strange to me tbh.

**Comment by bretthoerner** [comment]
> Only canonical names will be stored, but access is still possible via the legacy names in the entire codebase.

I don't have any review feedback, but I'm glad this came up... because this will effect Snuba and we need to update to be forwards compatible before this is merged since it consumes from Kafka and lives in its own codebase. If I'm reading correctly the interface keys in the JSON will actually be changed.

It should only take me a minute to update Snuba, but we probably need some way to ensure we inform downstream consumers (since we'll probably have more in the future).

**Comment by mitsuhiko** [comment]
@bretthoerner which are the downstream customers?

**Comment by bretthoerner** [comment]
Snuba is the only existing downstream consumer, but if the keys suddenly changed to (for example) `exception` our code would have stopped working and inserted `null`s into ClickHouse.

I just think we'll have more Kafka consumers in the future, but even if it's just Snuba... I don't think anyone knew to update our Kafka consumer for this.

**Comment by dcramer** [comment]
This is probably a good reason to say we lock in versions and dont change these keys in the future.

**Comment by mitsuhiko** [comment]
Why does snuba use the raw format and not the json export? I don’t think we
noticed anyone using the raw data dict. Where does that happen?

**Comment by bretthoerner** [comment]
Maybe I'm misunderstanding and the keys aren't changing?

We publish the `event.data.data` to Kafka and do a plain `json.loads` of whatever that is. Are you maintaining the old keys in there forever?

https://github.com/getsentry/getsentry/blob/5121257164ea45873cf271d5a80e61c63ca0237c/getsentry/processing.py#L55-L66

**Comment by mitsuhiko** [comment]
No. You are correct this will change. However more annoyingly this format
turns out to never haven been stable. All the rest uses the json export.

This is double annoying because this dependency was missed and we were
operating under the assumption that we are flexible in chaning this format.

I suppose changing to what data forwarding plugins do is not an option?

**Comment by jan-auer** [comment]
@bretthoerner You could use something like `CanonicalKeyDict` on your side, too.

**Comment by mitsuhiko** [comment]
@jan-auer there is absolutely no way to make the column rename work. As a result i had to go back to `event.data` being the `NodeField` itself. The tests pass but I did not actually verify that the data persisted is the correct one. I want to verify this tomorrow.

**Review by mattrobenolt — APPROVED** [review]
🤷‍♂️

**Pull request merged**
